File: memcontainers/kernel/rust/src/fs/servedfs.rs
//! Guest-served file servers — the deepest Plan 9 idea: a filesystem served by a
//! USER-SPACE guest. A guest calls `mc_sys_serve(path)` to become the server for
//! a subtree; the kernel mounts a `ServedFs` there. When another task opens a
//! path under the mount, the op is turned into a request the server guest
//! receives (`mc_sys_serve_recv`), handles, and answers (`mc_sys_serve_respond`).
//!
//! The cooperative dance is single-threaded: the requester's op enqueues a
//! request and returns `WouldBlock`, so the syscall yields. The server guest is
//! then scheduled, `recv`s the request, and `respond`s. On the requester's next
//! tick the op finds the response and completes. All state lives in linear
//! memory. The server can only ever return bytes to the requester — never a host
//! object.
//!
//! The same dance serves the whole terminal VFS surface, not just `open`: each
//! request carries an `op` (one of `abi::SERVE_OP_*`) so the server learns which
//! operation to perform. `open` returns the file content (then local `read`s
//! drain it — no per-read round trip); `stat` returns a 44-byte metadata record;
//! `readdir` returns typed directory entries; `mkdir`/`unlink`/`rename` return
//! status only. Namespace path resolution remains synchronous: unknown served
//! prefixes are treated as provisional searchable directories there, while
//! user-visible terminal `stat`/`lstat`/`chdir` use `stat_as` and may yield to
//! the server. Dedup is by caller, which makes a yield-retry idempotent (a
//! mutation is applied once, by the server, when it handles the recv).

use alloc::boxed::Box;
use alloc::collections::{BTreeMap, VecDeque};
use alloc::rc::Rc;
use alloc::string::String;
use alloc::vec::Vec;
use core::cell::RefCell;

use crate::wasm::abi::{
    SERVE_OP_MKDIR, SERVE_OP_OPEN, SERVE_OP_READDIR, SERVE_OP_RENAME, SERVE_OP_STAT,
    SERVE_OP_UNLINK,
};

use crate::fs::proxy;
use crate::vfs::traits::{
    CallerId, DirEntry, FileHandle, FileSystem, FsError, KPath, Metadata, OpenFlags, Result,
    SeekFrom,
};
use crate::wasm::abi::fs_result_from_errno;

/// A pending request routed to the server guest. `op` is one of `abi::SERVE_OP_*`;
/// `arg` is the secondary path (the `rename` target) and is empty for every
/// other op.
pub struct ServeRequest {
    pub id: u32,
    pub caller: CallerId,
    pub op: u32,
    pub path: String,
    pub arg: String,
}

/// The server guest's answer for one request. `status` is a WASI-style errno
/// (`0` = ok), decoded by [`fs_result_from_errno`]; `data` is op-specific
/// payload (file content for OPEN, stat record for STAT, typed names for
/// READDIR, empty for the mutating ops).
struct ServeResponse {
    status: i32,
    data: Vec<u8>,
}

/// The rendezvous between a `ServedFs` (mounted in the namespace) and the server
/// guest (which holds the same `Rc`). Shared, interior-mutable, single-threaded.
pub struct ServeChannel {
    next_id: u32,
    /// Requests waiting for the server to `recv`.
    requests: VecDeque<ServeRequest>,
    /// Answered requests waiting for the requester to pick up, by request id.
    responses: BTreeMap<u32, ServeResponse>,
    /// The in-flight request id per requester (so a yielding `open` re-poll does
    /// not enqueue twice).
    inflight: BTreeMap<CallerId, u32>,
    /// Metadata the kernel has learned from server answers. The mount root is
    /// always a directory; unknown paths stay provisional for resolution and are
    /// confirmed by terminal STAT.
    meta: BTreeMap<String, Metadata>,
    /// Set once the server guest exits; pending/new requests then fail (`EIO`).
    closed: bool,
}

impl ServeChannel {
    pub fn new() -> Rc<RefCell<ServeChannel>> {
        let mut meta = BTreeMap::new();
        meta.insert(String::from("/"), Metadata::dir());
        Rc::new(RefCell::new(ServeChannel {
            next_id: 1,
            requests: VecDeque::new(),
            responses: BTreeMap::new(),
            inflight: BTreeMap::new(),
            meta,
            closed: false,
        }))
    }

    /// Server side: take the next request to handle, if any.
    pub fn take_request(&mut self) -> Option<ServeRequest> {
        self.requests.pop_front()
    }

    /// Server side: inspect the next request without consuming it. Used so a
    /// too-small server buffer cannot drop a request and strand the caller.
    pub fn peek_request(&self) -> Option<&ServeRequest> {
        self.requests.front()
    }

    /// Server side: record the answer for `req_id`.
    pub fn respond(&mut self, req_id: u32, status: i32, data: Vec<u8>) -> bool {
        if !self.inflight.values().any(|&id| id == req_id) {
            return false;
        }
        self.responses
            .insert(req_id, ServeResponse { status, data });
        true
    }

    /// Server side: the server guest has exited — fail everything pending.
    pub fn close(&mut self) {
        self.closed = true;
    }

    fn cached_meta(&self, path: &str) -> Option<Metadata> {
        proxy::cached_meta(&self.meta, path)
    }

    fn remember_meta(&mut self, path: &str, meta: Metadata) {
        proxy::remember_meta(&mut self.meta, path, meta);
    }

    fn forget_path(&mut self, path: &str) {
        proxy::forget_path(&mut self.meta, path);
    }

    fn rename_path(&mut self, from: &str, to: &str) {
        proxy::rename_path(&mut self.meta, from, to);
    }

    fn forget_children_of(&mut self, dir: &str) {
        proxy::forget_children_of(&mut self.meta, dir);
    }
}

/// A filesystem whose operations are answered by a server guest.
pub struct ServedFs {
    channel: Rc<RefCell<ServeChannel>>,
}

impl ServedFs {
    pub fn new(channel: Rc<RefCell<ServeChannel>>) -> Self {
        ServedFs { channel }
    }

    /// The cooperative request/response dance every served op shares. On a
    /// caller's first touch it enqueues a `(op, path, arg)` request
    /// for the server guest and returns `WouldBlock`, so the syscall yields; on a
    /// later tick, once the server has answered, it returns that response (whose
    /// `status` the caller decodes). Dedup is by `caller`: a cooperative guest
    /// runs one syscall at a time, so at most one request per caller is ever in
    /// flight — which is exactly what makes the re-issue on a yield-retry
    /// idempotent (the server performs a mutation once, when it handles the
    /// recv; the caller's re-poll only collects the answer). A `WouldBlock` after
    /// the server has exited becomes `IoError` so no caller parks forever.
    fn request(&self, op: u32, path: &str, arg: &str, caller: CallerId) -> Result<ServeResponse> {
        let mut ch = self.channel.borrow_mut();
        if let Some(&rid) = ch.inflight.get(&caller) {
            if let Some(resp) = ch.responses.remove(&rid) {
                ch.inflight.remove(&caller);
                return Ok(resp);
            }
            if ch.closed {
                ch.inflight.remove(&caller);
                return Err(FsError::IoError);
            }
            return Err(FsError::WouldBlock);
        }
        if ch.closed {
            return Err(FsError::IoError);
        }
        // First touch: enqueue a request for the server, then yield.
        let id = ch.next_id;
        ch.next_id = ch.next_id.wrapping_add(1).max(1);
        ch.requests.push_back(ServeRequest {
            id,
            caller,
            op,
            path: String::from(path),
            arg: String::from(arg),
        });
        ch.inflight.insert(caller, id);
        Err(FsError::WouldBlock)
    }
}

impl FileSystem for ServedFs {
    fn open(
        &mut self,
        caller: CallerId,
        path: &KPath,
        _flags: OpenFlags,
    ) -> Result<Box<dyn FileHandle>> {
        let resp = self.request(SERVE_OP_OPEN, path.as_str(), "", caller)?;
        if let Err(e) = fs_result_from_errno(resp.status) {
            if e == FsError::NotFound {
                self.channel.borrow_mut().forget_path(path.as_str());
            }
            return Err(e);
        }
        self.channel
            .borrow_mut()
            .remember_meta(path.as_str(), Metadata::file(resp.data.len() as u64));
        // The server delivers the whole file at open time; reads drain it
        // locally (no per-read round trip).
        Ok(Box::new(ServedFileHandle {
            data: resp.data,
            offset: 0,
        }))
    }

    fn stat(&self, path: &KPath) -> Result<Metadata> {
        // Synchronous metadata is only for namespace path resolution and mode
        // checks. It must never yield. Cached server-proven metadata wins; an
        // unknown served path is a provisional searchable directory so direct
        // nested remote keys (`/mnt/s3/prefix/object`) can reach the terminal
        // served operation that asks the server for the real answer.
        if let Some(meta) = self.channel.borrow().cached_meta(path.as_str()) {
            return Ok(meta);
        }
        Ok(Metadata::dir())
    }

    fn stat_as(&self, caller: CallerId, path: &KPath) -> Result<Metadata> {
        let resp = self.request(SERVE_OP_STAT, path.as_str(), "", caller)?;
        if let Err(e) = fs_result_from_errno(resp.status) {
            if e == FsError::NotFound {
                self.channel.borrow_mut().forget_path(path.as_str());
            }
            return Err(e);
        }
        let meta = proxy::parse_metadata(&resp.data)?;
        self.channel
            .borrow_mut()
            .remember_meta(path.as_str(), meta.clone());
        Ok(meta)
    }

    fn readdir(&self, caller: CallerId, path: &KPath) -> Result<Vec<DirEntry>> {
        let resp = self.request(SERVE_OP_READDIR, path.as_str(), "", caller)?;
        if let Err(e) = fs_result_from_errno(resp.status) {
            if e == FsError::NotFound {
                self.channel.borrow_mut().forget_path(path.as_str());
            }
            return Err(e);
        }
        let parsed = proxy::parse_dirents(&resp.data, path.as_str())?;
        let mut entries = Vec::new();
        {
            let mut ch = self.channel.borrow_mut();
            ch.forget_children_of(path.as_str());
            for (entry, child, meta) in parsed {
                ch.remember_meta(&child, meta);
                entries.push(entry);
            }
        }
        Ok(entries)
    }

    fn mkdir(&mut self, caller: CallerId, path: &KPath) -> Result<()> {
        let resp = self.request(SERVE_OP_MKDIR, path.as_str(), "", caller)?;
        fs_result_from_errno(resp.status)?;
        self.channel
            .borrow_mut()
            .remember_meta(path.as_str(), Metadata::dir());
        Ok(())
    }

    fn unlink(&mut self, caller: CallerId, path: &KPath) -> Result<()> {
        let resp = self.request(SERVE_OP_UNLINK, path.as_str(), "", caller)?;
        fs_result_from_errno(resp.status)?;
        self.channel.borrow_mut().forget_path(path.as_str());
        Ok(())
    }

    fn rename(&mut self, caller: CallerId, from: &KPath, to: &KPath) -> Result<()> {
        let resp = self.request(SERVE_OP_RENAME, from.as_str(), to.as_str(), caller)?;
        fs_result_from_errno(resp.status)?;
        self.channel
            .borrow_mut()
            .rename_path(from.as_str(), to.as_str());
        Ok(())
    }
}

/// A handle over the content the server returned at open time. Reads drain the
/// buffer locally — no further server round trips.
struct ServedFileHandle {
    data: Vec<u8>,
    offset: usize,
}

impl FileHandle for ServedFileHandle {
    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        let start = self.offset.min(self.data.len());
        let n = (self.data.len() - start).min(buf.len());
        buf[..n].copy_from_slice(&self.data[start..start + n]);
        self.offset = start + n;
        Ok(n)
    }
    fn write(&mut self, _buf: &[u8]) -> Result<usize> {
        Err(FsError::PermissionDenied)
    }
    fn seek(&mut self, pos: SeekFrom) -> Result<u64> {
        let new = match pos {
            SeekFrom::Start(n) => n as i64,
            SeekFrom::Current(n) => self.offset as i64 + n,
            SeekFrom::End(n) => self.data.len() as i64 + n,
        };
        if new < 0 {
            return Err(FsError::InvalidPath);
        }
        self.offset = new as usize;
        Ok(self.offset as u64)
    }
    fn stat(&self) -> Result<Metadata> {
        Ok(Metadata::file(self.data.len() as u64))
    }
}

File: memcontainers/kernel/zig/src/fs/servedfs.zig
//! src/fs/servedfs.zig — guest-served filesystems (§2.5).
//!
//! Owns: request/response IDs, metadata cache behavior, and whole-file open behavior.
//! Invariants: A9, A7. ASYNC: pending marker, not deep park (§7.4).
//! Oracle (behavior to match): kernel/rust/src/fs/servedfs.rs.
//! Not here: the resident service registry (service/registry.zig).

const std = @import("std");
const constants = @import("constants_zig");
const vfs = @import("../vfs.zig");
const proxy = @import("proxy.zig");

const FsError = vfs.FsError;
const FileHandle = vfs.FileHandle;
const FileSystem = vfs.FileSystem;
const Metadata = vfs.Metadata;
const OpenFlags = vfs.OpenFlags;
const SeekFrom = vfs.SeekFrom;

pub const ServeRequest = struct {
    id: u32,
    caller: vfs.CallerId,
    op: u32,
    path: []u8,
    arg: []u8,

    pub fn deinit(self: *ServeRequest, gpa: std.mem.Allocator) void {
        gpa.free(self.path);
        gpa.free(self.arg);
        self.* = undefined;
    }
};

const ServeResponse = struct {
    status: i32,
    data: []u8,

    fn deinit(self: *ServeResponse, gpa: std.mem.Allocator) void {
        gpa.free(self.data);
    }
};

const ResponseEntry = struct {
    id: u32,
    response: ServeResponse,
};

const InflightEntry = struct {
    caller: vfs.CallerId,
    id: u32,
};

pub const ServeChannel = struct {
    gpa: std.mem.Allocator,
    refs: usize = 0,
    next_id: u32 = 1,
    requests: std.ArrayListUnmanaged(ServeRequest) = .empty,
    responses: std.ArrayListUnmanaged(ResponseEntry) = .empty,
    inflight: std.ArrayListUnmanaged(InflightEntry) = .empty,
    meta: std.StringHashMapUnmanaged(Metadata) = .{},
    closed: bool = false,

    pub fn create(gpa: std.mem.Allocator) *ServeChannel {
        const self = gpa.create(ServeChannel) catch @panic("OOM");
        self.* = .{ .gpa = gpa };
        proxy.rememberMeta(gpa, &self.meta, "/", Metadata.dir());
        return self;
    }

    pub fn retain(self: *ServeChannel) *ServeChannel {
        self.refs += 1;
        return self;
    }

    pub fn release(self: *ServeChannel) void {
        self.refs -|= 1;
        if (self.refs != 0) return;
        const gpa = self.gpa;
        self.deinit();
        gpa.destroy(self);
    }

    fn deinit(self: *ServeChannel) void {
        for (self.requests.items) |*r| r.deinit(self.gpa);
        self.requests.deinit(self.gpa);
        for (self.responses.items) |*r| r.response.deinit(self.gpa);
        self.responses.deinit(self.gpa);
        self.inflight.deinit(self.gpa);
        var it = self.meta.keyIterator();
        while (it.next()) |k| self.gpa.free(k.*);
        self.meta.deinit(self.gpa);
    }

    fn nextId(self: *ServeChannel) u32 {
        const id = self.next_id;
        self.next_id +%= 1;
        if (self.next_id == 0) self.next_id = 1;
        return id;
    }

    fn inflightIndex(self: *const ServeChannel, caller: vfs.CallerId) ?usize {
        for (self.inflight.items, 0..) |entry, i| {
            if (entry.caller == caller) return i;
        }
        return null;
    }

    fn responseIndex(self: *const ServeChannel, id: u32) ?usize {
        for (self.responses.items, 0..) |entry, i| {
            if (entry.id == id) return i;
        }
        return null;
    }

    pub fn peekRequest(self: *const ServeChannel) ?ServeRequest {
        if (self.requests.items.len == 0) return null;
        return self.requests.items[0];
    }

    pub fn takeRequest(self: *ServeChannel) ?ServeRequest {
        if (self.requests.items.len == 0) return null;
        return self.requests.orderedRemove(0);
    }

    pub fn respond(self: *ServeChannel, req_id: u32, status: i32, data: []const u8) bool {
        var found = false;
        for (self.inflight.items) |entry| {
            if (entry.id == req_id) {
                found = true;
                break;
            }
        }
        if (!found) return false;
        self.responses.append(self.gpa, .{
            .id = req_id,
            .response = .{
                .status = status,
                .data = self.gpa.dupe(u8, data) catch @panic("OOM"),
            },
        }) catch @panic("OOM");
        return true;
    }

    pub fn close(self: *ServeChannel) void {
        self.closed = true;
    }

    fn request(self: *ServeChannel, op: u32, path: []const u8, arg: []const u8, caller: vfs.CallerId) FsError!ServeResponse {
        if (self.inflightIndex(caller)) |idx| {
            const req_id = self.inflight.items[idx].id;
            if (self.responseIndex(req_id)) |ridx| {
                const resp = self.responses.orderedRemove(ridx).response;
                _ = self.inflight.orderedRemove(idx);
                return resp;
            }
            if (self.closed) {
                _ = self.inflight.orderedRemove(idx);
                return FsError.IoError;
            }
            return FsError.WouldBlock;
        }
        if (self.closed) return FsError.IoError;
        const id = self.nextId();
        self.requests.append(self.gpa, .{
            .id = id,
            .caller = caller,
            .op = op,
            .path = self.gpa.dupe(u8, path) catch @panic("OOM"),
            .arg = self.gpa.dupe(u8, arg) catch @panic("OOM"),
        }) catch @panic("OOM");
        self.inflight.append(self.gpa, .{ .caller = caller, .id = id }) catch @panic("OOM");
        return FsError.WouldBlock;
    }
};

pub const ServeOwner = struct {
    gpa: std.mem.Allocator,
    channel: *ServeChannel,

    pub fn create(gpa: std.mem.Allocator, channel: *ServeChannel) *ServeOwner {
        const self = gpa.create(ServeOwner) catch @panic("OOM");
        self.* = .{ .gpa = gpa, .channel = channel.retain() };
        return self;
    }

    pub fn release(self: *ServeOwner) void {
        self.channel.close();
        self.channel.release();
        self.gpa.destroy(self);
    }
};

pub const ServedFs = struct {
    gpa: std.mem.Allocator,
    channel: *ServeChannel,

    pub fn create(gpa: std.mem.Allocator, channel: *ServeChannel) *ServedFs {
        const self = gpa.create(ServedFs) catch @panic("OOM");
        self.* = .{ .gpa = gpa, .channel = channel.retain() };
        return self;
    }

    pub fn fileSystem(self: *ServedFs) FileSystem {
        return .{ .ptr = self, .vtable = &fs_vtable };
    }

    fn request(self: *ServedFs, op: u32, path: []const u8, arg: []const u8, caller: vfs.CallerId) FsError!ServeResponse {
        return self.channel.request(op, path, arg, caller);
    }

    fn open(self: *ServedFs, caller: vfs.CallerId, path: []const u8, _: OpenFlags) FsError!FileHandle {
        var resp = try self.request(constants.SERVE_OP_OPEN, path, "", caller);
        defer resp.deinit(self.gpa);
        try fsResultFromErrno(resp.status);
        proxy.rememberMeta(self.gpa, &self.channel.meta, path, Metadata.file(resp.data.len));
        const h = self.gpa.create(ServedFileHandle) catch @panic("OOM");
        h.* = .{ .gpa = self.gpa, .data = self.gpa.dupe(u8, resp.data) catch @panic("OOM") };
        return h.fileHandle();
    }

    fn stat(self: *ServedFs, path: []const u8) FsError!Metadata {
        return proxy.cachedMeta(&self.channel.meta, path) orelse Metadata.dir();
    }

    fn readdir(self: *ServedFs, caller: vfs.CallerId, path: []const u8, arena: std.mem.Allocator, out: *std.ArrayList(vfs.DirEntry)) FsError!void {
        var resp = try self.request(constants.SERVE_OP_READDIR, path, "", caller);
        defer resp.deinit(self.gpa);
        try fsResultFromErrno(resp.status);
        var parsed: std.ArrayList(proxy.ParsedDirent) = .empty;
        try proxy.parseDirents(arena, resp.data, path, &parsed);
        proxy.forgetChildrenOf(self.gpa, &self.channel.meta, path);
        for (parsed.items) |p| {
            proxy.rememberMeta(self.gpa, &self.channel.meta, p.child, p.metadata);
            out.append(arena, p.entry) catch @panic("OOM");
        }
    }

    fn mkdir(self: *ServedFs, caller: vfs.CallerId, path: []const u8) FsError!void {
        var resp = try self.request(constants.SERVE_OP_MKDIR, path, "", caller);
        defer resp.deinit(self.gpa);
        try fsResultFromErrno(resp.status);
        proxy.rememberMeta(self.gpa, &self.channel.meta, path, Metadata.dir());
    }

    fn unlink(self: *ServedFs, caller: vfs.CallerId, path: []const u8) FsError!void {
        var resp = try self.request(constants.SERVE_OP_UNLINK, path, "", caller);
        defer resp.deinit(self.gpa);
        try fsResultFromErrno(resp.status);
        proxy.forgetPath(self.gpa, &self.channel.meta, path);
    }

    fn rename(self: *ServedFs, caller: vfs.CallerId, from: []const u8, to: []const u8) FsError!void {
        var resp = try self.request(constants.SERVE_OP_RENAME, from, to, caller);
        defer resp.deinit(self.gpa);
        try fsResultFromErrno(resp.status);
        proxy.renamePath(self.gpa, &self.channel.meta, from, to);
    }

    const fs_vtable = FileSystem.VTable{
        .open = fsOpen,
        .stat = fsStat,
        .readdir = fsReaddir,
        .mkdir = fsMkdir,
        .unlink = fsUnlink,
        .rename = fsRename,
        .symlink = vfs.fsSymlinkUnsupported,
        .link = vfs.fsLinkUnsupported,
        .readlink = vfs.fsReadlinkUnsupported,
        .setMode = vfs.fsSetModeUnsupported,
        .setTimes = vfs.fsSetTimesUnsupported,
    };

    fn self_(p: *anyopaque) *ServedFs {
        return @ptrCast(@alignCast(p));
    }
    fn fsOpen(p: *anyopaque, caller: vfs.CallerId, path: []const u8, flags: OpenFlags) FsError!FileHandle {
        return self_(p).open(caller, path, flags);
    }
    fn fsStat(p: *anyopaque, path: []const u8) FsError!Metadata {
        return self_(p).stat(path);
    }
    fn fsReaddir(p: *anyopaque, caller: vfs.CallerId, path: []const u8, arena: std.mem.Allocator, out: *std.ArrayList(vfs.DirEntry)) FsError!void {
        return self_(p).readdir(caller, path, arena, out);
    }
    fn fsMkdir(p: *anyopaque, caller: vfs.CallerId, path: []const u8) FsError!void {
        return self_(p).mkdir(caller, path);
    }
    fn fsUnlink(p: *anyopaque, caller: vfs.CallerId, path: []const u8) FsError!void {
        return self_(p).unlink(caller, path);
    }
    fn fsRename(p: *anyopaque, caller: vfs.CallerId, from: []const u8, to: []const u8) FsError!void {
        return self_(p).rename(caller, from, to);
    }
};

fn fsResultFromErrno(errno: i32) FsError!void {
    if (errno == 0) return;
    return switch (errno) {
        constants.ENOENT => FsError.NotFound,
        constants.EEXIST => FsError.AlreadyExists,
        constants.ENOTDIR => FsError.NotDir,
        constants.EISDIR => FsError.IsDir,
        constants.EPERM => FsError.PermissionDenied,
        constants.EACCES => FsError.AccessDenied,
        constants.EINVAL => FsError.InvalidPath,
        constants.ENOTEMPTY => FsError.NotEmpty,
        constants.EBADF => FsError.BadFileDescriptor,
        constants.ENOSYS => FsError.NotImplemented,
        constants.EXDEV => FsError.CrossDevice,
        constants.EAGAIN => FsError.WouldBlock,
        constants.EMSGSIZE => FsError.MessageTooBig,
        constants.ELOOP => FsError.Loop,
        else => FsError.IoError,
    };
}

const ServedFileHandle = struct {
    gpa: std.mem.Allocator,
    data: []u8,
    offset: usize = 0,

    fn fileHandle(self: *ServedFileHandle) FileHandle {
        return .{ .ptr = self, .vtable = &handle_vtable };
    }

    fn read(self: *ServedFileHandle, out: []u8) FsError!usize {
        const start = @min(self.offset, self.data.len);
        const n = @min(out.len, self.data.len - start);
        if (n != 0) @memcpy(out[0..n], self.data[start .. start + n]);
        self.offset = start + n;
        return n;
    }

    fn write(_: *ServedFileHandle, _: []const u8) FsError!usize {
        return FsError.PermissionDenied;
    }

    fn seek(self: *ServedFileHandle, pos: SeekFrom) FsError!u64 {
        const next = switch (pos) {
            .start => |n| @as(i64, @intCast(n)),
            .current => |n| @as(i64, @intCast(self.offset)) + n,
            .end => |n| @as(i64, @intCast(self.data.len)) + n,
        };
        if (next < 0) return FsError.InvalidPath;
        self.offset = @intCast(next);
        return @intCast(self.offset);
    }

    fn stat(self: *ServedFileHandle) FsError!Metadata {
        return Metadata.file(self.data.len);
    }

    fn truncate(_: *ServedFileHandle, _: u64) FsError!void {
        return FsError.PermissionDenied;
    }

    fn close(self: *ServedFileHandle) void {
        const gpa = self.gpa;
        gpa.free(self.data);
        gpa.destroy(self);
    }

    const handle_vtable = FileHandle.VTable{
        .read = hRead,
        .write = hWrite,
        .seek = hSeek,
        .stat = hStat,
        .truncate = hTruncate,
        .close = hClose,
    };

    fn self_(p: *anyopaque) *ServedFileHandle {
        return @ptrCast(@alignCast(p));
    }
    fn hRead(p: *anyopaque, out: []u8) FsError!usize {
        return self_(p).read(out);
    }
    fn hWrite(p: *anyopaque, bytes: []const u8) FsError!usize {
        return self_(p).write(bytes);
    }
    fn hSeek(p: *anyopaque, pos: SeekFrom) FsError!u64 {
        return self_(p).seek(pos);
    }
    fn hStat(p: *anyopaque) FsError!Metadata {
        return self_(p).stat();
    }
    fn hTruncate(p: *anyopaque, size: u64) FsError!void {
        return self_(p).truncate(size);
    }
    fn hClose(p: *anyopaque) void {
        self_(p).close();
    }
};
